rholang/src/main/scala/rholang/rosette/Equivalence.scala

```diff
@@ -0,0 +1,330 @@
+        allCPatternEquivalent(
+          env1, contr1.listcpattern_.asScala.toList,
+          env2, contr2.listcpattern_.asScala.toList
+          ) match {
+            case None => false
+            case Some((newenv1,newenv2)) =>
+              env1.equivalent(contr1.var_, env2, contr2.var_) &&
+                structurallyEquivalent(newenv1, contr1.proc_, newenv2, contr2.proc_)
+          }
+      }
+      case (p1: PPar, p2: PPar) => parEquiv(parLeaves(p1),parLeaves(p2))
+      case _ => false
+    }
+
+  def bindEquivalent(env1: DeBruijn, b1: Bind, env2: DeBruijn, b2: Bind): Option[(CPattern,CPattern)] =
+    (b1, b2) match {
+      case (inpBind1: InputBind, inpBind2: InputBind) =>
+        if (nameEquivalent(env1, inpBind1.chan_, env2, inpBind2.chan_)) {
+          Some((inpBind1.cpattern_, inpBind2.cpattern_))
+        } else None
+      case (condInpBind1: CondInputBind, condInpBind2: CondInputBind) =>
+        val check =
+          nameEquivalent(env1, condInpBind1.chan_, env2, condInpBind2.chan_) &&
+          structurallyEquivalent(env1, condInpBind1.proc_, env2, condInpBind2.proc_)
+        if (check) {
+          Some((condInpBind1.cpattern_, condInpBind2.cpattern_))
+        } else None
+      case _ => None 
+    }
+
+  def allBindsEquivalent(env1: DeBruijn, b1: List[Bind], env2: DeBruijn, b2: List[Bind]): Option[(List[CPattern],List[CPattern])] =
+    b1 match {
+      case Nil => b2 match {
+        case Nil => Some((Nil,Nil))
+        case _ => None
+      }
+      case head :: tail => b2.partition {
+        bind => !(bindEquivalent(env1,bind,env2,head).isEmpty)
+      } match {
+        case (Nil,_) => None
+        case (eqhd,eqtl) =>
+          val init = (
+            false,
+            List[CPattern](),
+            List[CPattern](),
+            List[Bind](),
+            eqhd.tail
+          )
+          eqhd.foldLeft(init) {
+            (rejects, bnd) => rejects match {
+              case (false, names1, names2, r, l) =>
+                allBindsEquivalent(env1, r ++ l ++ eqtl, env2, tail) match {
+                  case Some((name1,name2)) => (
+                    true,
+                    name1 ++ names1,
+                    name2 ++ names2,
+                    r, l
+                  )
+                  case None => l match {
+                    case Nil => (
+                      false, names1, names2,
+                      r ++ List(bnd),
+                      Nil
+                    )
+                    case x :: xs1 => (
+                      false, names1, names2,
+                      r ++ List(bnd),
+                      xs1
+                    )
+                  }
+                }
+              case tuple => tuple
+            }
+          } match {
+            case (false,_,_,_,_) => None
+            case (true,names1,names2,_,_) => Some((names1,names2))
+          }
+        }
+    }
+
+  def alphaEquivalent(p1: Proc, p2: Proc): Boolean = ???
```

Comment:
You don't need this, that's what the environments are for. (If the vars represent the same Debruijn level, they are alpha equivalent)
Substitution is easier in rholang than rho calculus because we have patterns and don't use names as variables.